src/main/java/org/elasticsearch/index/engine/ShadowEngine.java

```diff
@@ -139,12 +143,24 @@ public CommitId flush(boolean force, boolean waitIfOngoing) throws EngineExcepti
          * dec the store reference which can essentially close the store and unless we can inc the reference
          * we can't use it.
          */
-        CommitId id = null;
+        CommitId id = new CommitId(new BytesRef()); // empty by default
         store.incRef();
         try (ReleasableLock lock = readLock.acquire()) {
             // reread the last committed segment infos
-            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
-            id = CommitId.readCommitID(store, lastCommittedSegmentInfos);
+            while(true) {
+                // reread the last committed segment infos
+                lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+                try {
+                    id = CommitId.readCommitID(store, lastCommittedSegmentInfos);
+                    break;
+                } catch (FileNotFoundException | NoSuchFileException ex) {
+                    if (IndexMetaData.isOnSharedFilesystem(config().getIndexSettings()) == false) {
+                        throw ex;
+                    }
+                    // we have to read this in a loop since there is a potential race-condition between reading
+                    // the segment info and reading the commit ID/
+                }
```

Comment:
I think doing it in a `FindSegmentsFile` block would be better, but there is still going to be a race condition on actually reading the segments file, so it will still need to be in a loop?